pkg/tsdb/influxdb/influxql/converter/converter.go

```diff
@@ -0,0 +1,395 @@
+		if err != nil {
+			return rspErr(err)
+		}
+		for l1Field, err := iter.ReadObject(); l1Field != ""; l1Field, err = iter.ReadObject() {
+			if err != nil {
+				return rspErr(err)
+			}
+			switch l1Field {
+			case "series":
+				rsp = readSeries(iter, frameName, query)
+			case "":
+				break l1Fields
+			default:
+				_, err := iter.Read()
+				if err != nil {
+					return rspErr(err)
+				}
+			}
+		}
+	}
+
+	return rsp
+}
+
+func readSeries(iter *jsonitere.Iterator, frameName []byte, query *models.Query) *backend.DataResponse {
+	var measurement string
+	var tags map[string]string
+	var columns []string
+	var valueFields data.Fields
+	var hasTimeColumn bool
+	rsp := &backend.DataResponse{Frames: make(data.Frames, 0)}
+	for more, err := iter.ReadArray(); more; more, err = iter.ReadArray() {
+		if err != nil {
+			return rspErr(err)
+		}
+
+		for l1Field, err := iter.ReadObject(); l1Field != ""; l1Field, err = iter.ReadObject() {
+			if err != nil {
+				return rspErr(err)
+			}
+			switch l1Field {
+			case "name":
+				if measurement, err = iter.ReadString(); err != nil {
+					return rspErr(err)
+				}
+			case "tags":
+				if tags, err = readTags(iter); err != nil {
+					return rspErr(err)
+				}
+			case "columns":
+				columns, err = readColumns(iter)
+				if err != nil {
+					return rspErr(err)
+				}
+				if columns[0] == "time" {
+					hasTimeColumn = true
+				}
+			case "values":
+				valueFields, err = readValues(iter, hasTimeColumn)
+				if err != nil {
+					return rspErr(err)
+				}
+			default:
+				v, err := iter.Read()
+				if err != nil {
+					return rspErr(err)
+				}
+				fmt.Printf("unsupported key: %s / %v\n", l1Field, v)
+			}
+		}
+
+		if util.GetVisType(query.ResultFormat) == util.TableVisType {
+			handleTableFormatFirstFrame(rsp, measurement, query)
+			handleTableFormatFirstField(rsp, valueFields, columns)
+			handleTableFormatTagFields(rsp, valueFields, tags)
+			handleTableFormatValueFields(rsp, valueFields, tags, columns)
+		} else {
+			// time_series response format
+			if hasTimeColumn {
+				// Frame with time column
+				newFrames := handleTimeSeriesFormatWithTimeColumn(valueFields, tags, columns, measurement, frameName, query)
+				rsp.Frames = append(rsp.Frames, newFrames...)
+			} else {
+				// Frame without time column
+				newFrame := handleTimeSeriesFormatWithoutTimeColumn(valueFields, columns, measurement, query)
+				rsp.Frames = append(rsp.Frames, newFrame)
+			}
+		}
+	}
+
+	return rsp
+}
+
+func readTags(iter *jsonitere.Iterator) (map[string]string, error) {
+	tags := make(map[string]string)
+	for l1Field, err := iter.ReadObject(); l1Field != ""; l1Field, err = iter.ReadObject() {
+		if err != nil {
+			return nil, err
+		}
+		value, err := iter.ReadString()
+		if err != nil {
+			return nil, err
+		}
+		tags[l1Field] = value
+	}
+	return tags, nil
+}
+
+func readColumns(iter *jsonitere.Iterator) (columns []string, err error) {
+	columns = make([]string, 0)
```

Comment:
Is this necessary? `columns` is already appendable. I would use an explicit return instead of a naked one for clarity: `return columns`.